crypto/stack/stack.c

```diff
@@ -21,10 +28,43 @@ struct stack_st {
     OPENSSL_sk_compfunc comp;
 };
 
-#undef MIN_NODES
-#define MIN_NODES       4
+/*
+ * Calculate the array growth based on the target size.
+ *
+ * The growth faction is a rational number and is defined by a numerator
+ * and a denominator.  According to Andrew Koenig in his paper "Why Are
+ * Vectors Efficient?" from JOOP 11(5) 1998, this factor should be less
+ * than the golden ratio (1.618...).
+ *
+ * We use 3/2 = 1.5 for simplicty of calculation and overflow checking.
+ * Another option 8/5 = 1.6 allows for slightly faster growth, although safe
+ * computation is more difficult.
+ *
+ * The limit to avoid overflow is spot on.  2^n-1 is congruent to 0 mod 3
+ * when n is even and 1 mod 3 when n is odd.  When n is odd, int_max = 2^n-1
+ * produces limit = 2*floor(int_max/3) which is thus low by one, hence the
+ * correction term that adds one back for this case.
+ */
+static ossl_inline size_t compute_growth(size_t target, size_t current)
+{
+    const size_t limit = (INT_MAX / 3) * 2 + INT_MAX % 3;
+    size_t sz, n = current;
+
+    while (n <= target) {
+        /* This shouldn't be possible but check it anyway */
+        if (n >= INT_MAX)
+            return 0;
 
-#include <errno.h>
+        /* Expand the size by a factor of 3/2 if it is within range */
+        sz = n < limit ? n + n / 2 : INT_MAX;
+
+        /* Overflow check, which is possible if SIZE_MAX < INT_MAX */
+        if (sz < n)
```

Comment:
Well, I see contradiction here. If you consider SIZE_MAX < INT_MAX as possibility, then you can't calculate limit the way you do. I mean there is implicit cast to narrower type, so which value would *that* yield? If we make such assumption, then limit should be two thirds of *minimum* between SIZE_MAX/sizeof(void *) and INT_MAX. Note that it incidentally works on ILP32 systems... But note that we don't have to make such assumption, or rather we can *choose* not to make one by declaring that ILP32 is the minimum supported. [Well, formally speaking one can still argue that size_t can be narrower than 32 bits, because there is no S in ILP.] Myself I'm fine  with *not* making the assumption.

Reply:
Good catch.

Are you suggesting to set the limit to the minimum?
Or that the SIZE_MAX >= INT_MAX assumption is reasonable?  Like you, I'm not aware of a machine that doesn't have this.

How about defining a constant equal to the minimum of SIZE_MAX/sizeof(char *) and INT_MAX and using this instead of both SIZE_MAX and INT_MAX for the limit checking?

Reply:
> Are you suggesting to set the limit to the minimum?
> Or that the SIZE_MAX >= INT_MAX assumption is reasonable?

Well, what I'm saying is that *IF* we make the assumption that SIZE_MAX can be less than INT_MAX, then we wouldn't have another choice. Then I'm suggesting *not* to make the assumption, because it's reasonable in the landscape of supported configurations(\*). Then I'm saying that minimum(\*\*) might be sensible *anyway* for ILP32 account. In other words I kind of suggest both: set to minimum(\*\*), but without making the assumption. I.e. it's not "or", but "and".

BTW, not making the assumption means that this `if (sz < n)` would have to be omitted. Basically making or not making the assumption is also question of readability. As discussed earlier, readability, among other things, is about how many questions pops in your mind when you read the code, and less is better. What I'm hinting at is that more assumptions lead to less readable code and this `if` is the illustration. I mean when you see it you have to ask yourself "why" and seek for answer [which would be "not applicable in this case" for vast majority].

(\*) Not to mention that it's likely to be made all over anyway. It might be appropriate to reflect the fact that we expect SIZE_MAX to be not less than INT_MAX in test/sanitytest.c, so that we would know when it breaks. [And if it does, we would still have an option to disclaim support, as opposite to fixing the whole thing.]

(\*\*) minimum between SIZE_MAX/sizeof(void *) and INT_MAX, not SIZE_MAX and INT_MAX.

Reply:
> set to minimum, but without making the assumption

Just in case, reason for that is kind of readability. We have to check for SIZE_MAX/sizeof(void*) in either case, don't we? And I'd say that it would be more appropriate if all checks are performed in one place, instead of two. I mean now compute_growth checks for INT_MAX, while SIZE_MAX/sizeof(void*) boundary is enforced elsewhere...

Reply:
Using a single limit value simplified the range checking quite a bit.
I've moved the limit calculation code down next to the reserve function so the range checks are in one place.